### PR TITLE
add battery health info

### PR DESCRIPTION
**Pull request by pkowalsk1**
bump::patch

The capability to publish battery health information in the BatteryState message has been implemented.

### CODE REVIEW

File: panther_battery/src/adc_node.py [hunk 1]
```diff
@@ -186,6 +193,25 @@ def _publish_battery_msg(
         else:
             battery_msg.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_DISCHARGING
 
+        # check battery health
+        with self._lock:
+            if V_bat < self.V_BAT_FATAL_MIN:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_DEAD
+                rospy.logerr(f'[{rospy.get_name()}] battery voltage critically low!')
```
Comment (Kotochleb) [inline panther_battery/src/adc_node.py]: This should throttled

File: panther_battery/src/adc_node.py [hunk 2]
```diff
@@ -186,6 +193,25 @@ def _publish_battery_msg(
...
+            elif battery_msg.percentage > 1.1:
```
Comment (Kotochleb) [inline panther_battery/src/adc_node.py]: I would rather look at voltage. If the value exceeds 45 volts, this should mean overvoltage. By this should be after moving average

File: panther_battery/src/roboteq_republisher_node.py [hunk 3]
```diff
@@ -71,6 +73,13 @@ def _battery_pub_timer_cb(self, *args) -> None:
                 battery_msg.present = True
                 # TODO:
                 # battery_msg.power_supply_health
+                            
+                if self._battery_voltage < self.V_BAT_FATAL_MIN:
+                    battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_DEAD
+                elif battery_msg.percentage > 1.1:
```
Comment (Kotochleb) [inline panther_battery/src/roboteq_republisher_node.py]: Same as in other node

**Review by Kotochleb — CHANGES_REQUESTED** [review]

File: panther_battery/src/adc_node.py [hunk 4]
```diff
@@ -186,8 +205,42 @@ def _publish_battery_msg(
...
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_GOOD
+
+        if error_msg is not None:
+            rospy.logerr_throttle_identical(10.0, f'[{rospy.get_name()}] {error_msg}')
+
         bat_pub.publish(battery_msg)
 
+    def _count_volt_mean(self, label: Union[rospy.Publisher, str], new_val: float) -> float:
+        self._V_bat_mean[label] += (
+            -self._V_bat_hist[label][0] / self._volt_mean_length + new_val / self._volt_mean_length
```
Comment (Kotochleb) [inline panther_battery/src/adc_node.py]: A comment would be nice. I had trouble understanding what is going on. Anyway, cool approach to the problem

File: panther_battery/src/adc_node.py [hunk 5]
```diff
@@ -186,8 +205,42 @@ def _publish_battery_msg(
...
+            error_msg = None
+            if V_bat_mean < self.V_BAT_FATAL_MIN:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_DEAD
+                error_msg = 'Battery voltage is critically low!'
+            elif V_bat_mean > self.V_BAT_FATAL_MAX:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_OVERVOLTAGE
+                error_msg = 'Battery overvoltage!'
+            elif temp_bat >= self._high_bat_temp:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_OVERHEAT
+                error_msg = 'Battery temperature is dangerously high!'
```
Comment (Kotochleb) [inline panther_battery/src/adc_node.py]:
```suggestion
                error_msg = 'Battery is overheating!'
```
For some reason, "dangerously" sounds strange in this case. IDK

**Review by Kotochleb — CHANGES_REQUESTED** [review]
For some reason I feel like `POWER_SUPPLY_HEALTH_DEAD` and `POWER_SUPPLY_HEALTH_OVERHEAT` could yield `logfatal` instead of `logerr`, but this would look worse in code

**Comment by pkowalsk1** [comment]
> For some reason I feel like `POWER_SUPPLY_HEALTH_DEAD` and `POWER_SUPPLY_HEALTH_OVERHEAT` could yield `logfatal` instead of `logerr`, but this would look worse in code

Our products never encounter fatal errors ;)

File: panther_battery/src/roboteq_republisher_node.py [hunk 6]
```diff
@@ -55,25 +67,48 @@ def _battery_pub_timer_cb(self, *args) -> None:
...
+                if error_msg is not None:
+                    rospy.logerr_throttle_identical(10.0, f'[{rospy.get_name()}] {error_msg}')
 
             self._battery_pub.publish(battery_msg)
 
+    def _update_volt_mean(self, new_val: float) -> float:
+        self._battery_voltage_mean += (
+            -self._battery_voltage_hist[0] / self._volt_mean_length
+            + new_val / self._volt_mean_length
+        )
```
Comment (Kotochleb) [inline panther_battery/src/roboteq_republisher_node.py]: This node could have the comment as well

**Review by Kotochleb — CHANGES_REQUESTED** [review]

**Review by Kotochleb — APPROVED** [review]

**Pull request merged**